File: 42_Gunner/python/gunner.py
#!/usr/bin/env python3
#
# Ported to Python by @iamtraction

from math import sin
from random import random


def gunner() -> None:
    gun_range = int(40000 * random() + 20000)

    print("\nMAXIMUM RANGE OF YOUR GUN IS", gun_range, "YARDS.")

    killed_enemies = 0
    S1 = 0

    while True:
        target_distance = int(gun_range * (0.1 + 0.8 * random()))
        shots = 0

        print("\nDISTANCE TO THE TARGET IS", target_distance, "YARDS.")

        while True:
            elevation = float(input("\n\nELEVATION? "))

            if elevation > 89:
                print("MAXIMUM ELEVATION IS 89 DEGREES.")
                continue

            if elevation < 1:
                print("MINIMUM ELEVATION IS ONE DEGREE.")
                continue

            shots += 1

            if shots < 6:
                B2 = 2 * elevation / 57.3
                shot_impact = gun_range * sin(B2)
                shot_proximity = target_distance - shot_impact
                shot_proximity_int = int(shot_proximity)

                if abs(shot_proximity_int) < 100:
                    print(
                        "*** TARGET DESTROYED *** ",
                        shots,
                        "ROUNDS OF AMMUNITION EXPENDED.",
                    )
                    S1 += shots
                    if killed_enemies == 4:
                        print("\n\nTOTAL ROUNDS EXPENDED WERE: ", S1)
                        if S1 > 18:
                            print("BETTER GO BACK TO FORT SILL FOR REFRESHER TRAINING!")
                            return
                        else:
                            print("NICE SHOOTING !!")
                            return
                    else:
                        killed_enemies += 1
                        print(
                            "\nTHE FORWARD OBSERVER HAS SIGHTED MORE ENEMY ACTIVITY..."
                        )
                        break
                else:
                    if shot_proximity_int > 100:
                        print("SHORT OF TARGET BY", abs(shot_proximity_int), "YARDS.")
                    else:
                        print("OVER TARGET BY", abs(shot_proximity_int), "YARDS.")
            else:
                print("\nBOOM !!!!   YOU HAVE JUST BEEN DESTROYED BY THE ENEMY.\n\n\n")
                print("BETTER GO BACK TO FORT SILL FOR REFRESHER TRAINING!")
                return


if __name__ == "__main__":
    print(" " * 33 + "GUNNER")
    print(" " * 15 + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY")

    print("\n\n\n")

    print("YOU ARE THE OFFICER-IN-CHARGE, GIVING ORDERS TO A GUN")
    print("CREW, TELLING THEM THE DEGREES OF ELEVATION YOU ESTIMATE")
    print("WILL PLACE A PROJECTILE ON TARGET.  A HIT WITHIN 100 YARDS")
    print("OF THE TARGET WILL DESTROY IT.")

    while True:
        gunner()

        Y = input("TRY AGAIN (Y OR N)? ")
        if Y != "Y":
            print("\nOK.  RETURN TO BASE CAMP.")
            break

File: 42_Gunner/perl/gunner.pl
#!/usr/bin/perl

# Gunner program in Perl
#   Required extensive restructuring to remove all of the GOTO's.
# Translated by Kevin Brannen (kbrannen)

use strict;
use warnings;

# globals
my $Max_range = int(40000*rand(1)+20000);
my $Total_shots = 0;
my $Games = 0;

print "\n";
print " " x 30, "GUNNER\n";
print " " x 15, "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n\n\n";

print "YOU ARE THE OFFICER-IN-CHARGE, GIVING ORDERS TO A GUN\n";
print "CREW, TELLING THEM THE DEGREES OF ELEVATION YOU ESTIMATE\n";
print "WILL PLACE A PROJECTILE ON TARGET.  A HIT WITHIN 100 YARDS\n";
print "OF THE TARGET WILL DESTROY IT.\n\n";
print "MAXIMUM RANGE OF YOUR GUN IS $Max_range YARDS.\n\n";

GAME: while (1)
{
    my $target_dist = int($Max_range * (.1 + .8 * rand(1)));
    my $shots = 0;
    print "DISTANCE TO THE TARGET IS $target_dist YARDS.\n\n";
    while (1)
    {
        my $elevation = get_elevation(); # in degrees
        $shots++;
        my $dist = int($target_dist - ($Max_range * sin(2 * $elevation / 57.3)));
        if (abs($dist) < 100)
        {
            print "*** TARGET DESTROYED ***  $shots ROUNDS OF AMMUNITION EXPENDED.\n";
            $Total_shots += $shots;
            if ($Games++ == 4)
            {
                print "\n\nTOTAL ROUNDS EXPENDED WERE: $Total_shots\n";
                if ($Total_shots > 18) { print "BETTER GO BACK TO FORT SILL FOR REFRESHER TRAINING!\n"; }
                else                   { print "NICE SHOOTING !!\n"; }
                last;
            }
            print "\nTHE FORWARD OBSERVER HAS SIGHTED MORE ENEMY ACTIVITY...\n";
            next GAME;
        }
        if ($dist > 100) { print "SHORT OF TARGET BY ", abs($dist)," YARDS.\n"; }
        else             { print "OVER TARGET BY ", abs($dist), " YARDS.\n"; }

        if ($shots >= 5)
        {
            print "\nBOOM !!!!   YOU HAVE JUST BEEN DESTROYED BY THE ENEMY.\n\n\n\n"; 
            print "BETTER GO BACK TO FORT SILL FOR REFRESHER TRAINING!\n";
            last;
        }
    }

    print "\nTRY AGAIN (Y OR N): ";
    chomp(my $ans=uc(<>));
    if ($ans ne "Y") { last; }
    else             { $Games = 0; $Total_shots = 0; }
}

print "\nOK.  RETURN TO BASE CAMP.\n";

####################################

sub get_elevation
{
    my $elevation;
    while (1)
    {
        print "\nELEVATION: ";
        chomp($elevation = <>);
        if    ($elevation > 89) { print "MAXIMUM ELEVATION IS 89 DEGREES.\n"; }
        elsif ($elevation < 1)  { print "MINIMUM ELEVATION IS ONE DEGREE.\n"; }
        else                    { return $elevation; }
    }
}
